src/index/sparse/sparse_index_node.cc

```diff
@@ -0,0 +1,219 @@
+
+ public:
+    explicit SparseInvertedIndexNode(const int32_t& /*version*/, const Object& /*object*/)
+        : search_pool_(ThreadPool::GetGlobalSearchThreadPool()) {
+    }
+
+    ~SparseInvertedIndexNode() override {
+        if (index_ != nullptr) {
+            delete index_;
+            index_ = nullptr;
+        }
+    }
+
+    Status
+    Train(const DataSet& dataset, const Config& config) override {
+        auto cfg = static_cast<const SparseInvertedIndexConfig&>(config);
+        if (!IsMetricType(cfg.metric_type.value(), metric::IP)) {
+            LOG_KNOWHERE_ERROR_ << Type() << " only support metric_type: IP";
+            return Status::invalid_metric_type;
+        }
+        auto drop_ratio_build = cfg.drop_ratio_build.value_or(0.0f);
+        auto index = new sparse::InvertedIndex<T>();
+        index->SetUseWand(use_wand);
+        index->Train(static_cast<const sparse::SparseRow<T>*>(dataset.GetTensor()), dataset.GetRows(),
+                     drop_ratio_build);
+        if (index_ != nullptr) {
+            LOG_KNOWHERE_WARNING_ << Type() << " deleting old index during train";
+            delete index_;
+        }
+        index_ = index;
+        return Status::success;
+    }
+
+    Status
+    Add(const DataSet& dataset, const Config& config) override {
+        if (!index_) {
+            LOG_KNOWHERE_ERROR_ << "Could not add data to empty " << Type();
+            return Status::empty_index;
+        }
+        return index_->Add(static_cast<const sparse::SparseRow<T>*>(dataset.GetTensor()), dataset.GetRows(),
+                           dataset.GetDim());
+    }
+
+    [[nodiscard]] expected<DataSetPtr>
+    Search(const DataSet& dataset, const Config& config, const BitsetView& bitset) const override {
+        if (!index_) {
+            LOG_KNOWHERE_ERROR_ << "Could not search empty " << Type();
+            return expected<DataSetPtr>::Err(Status::empty_index, "index not loaded");
+        }
+        auto cfg = static_cast<const SparseInvertedIndexConfig&>(config);
+        auto nq = dataset.GetRows();
+        auto queries = static_cast<const sparse::SparseRow<T>*>(dataset.GetTensor());
+        auto k = cfg.k.value();
+        auto refine_factor = cfg.refine_factor.value_or(10);
+        auto drop_ratio_search = cfg.drop_ratio_search.value_or(0.0f);
+
+        auto p_id = new sparse::label_t[nq * k];
+        auto p_dist = new float[nq * k];
+
+        std::vector<folly::Future<folly::Unit>> futs;
+        futs.reserve(nq);
+        for (int64_t idx = 0; idx < nq; ++idx) {
+            futs.emplace_back(search_pool_->push([&, idx = idx]() {
+                index_->Search(queries[idx], k, drop_ratio_search, p_dist + idx * k, p_id + idx * k, refine_factor,
+                               bitset);
+            }));
+        }
+        for (auto& fut : futs) {
+            fut.wait();
+        }
+        return GenResultDataSet(nq, k, p_id, p_dist);
```

Comment:
`for (auto& fut : futs) RETURN_IF_ERROR(fut.result().value());` is missing

Reply:
changed to `fut.get()`

Reply:
I've added comments to #355 . Basically, it does not work as desired.